godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Color.kt

```diff
@@ -0,0 +1,806 @@
+                    else -> return -1.0
+                }
+
+                ig += if (i == 0)
+                    v * 16
+                else
+                    v
+            }
+            return ig
+        }
+    }
+
+    //CONSTRUCTOR
+
+    constructor() : this(0.0, 0.0, 0.0, 1.0)
+
+    constructor(r: Number, g: Number, b: Number, a: Number = 1.0) :
+        this(r.toRealT(), g.toRealT(), b.toRealT(), a.toRealT())
+
+    constructor(from: String) : this() {
+        var color = from
+        if (color.isEmpty())
+            return
+        if (color[0] == '#')
+            color = color.substring(1, color.length - 1)
+
+        val alpha = when {
+            color.length == 8 -> true
+            color.length == 6 -> false
+            else -> return
+        }
+
+        var a = 255
+        if (alpha) {
+            a = parseCol(color, 0).toInt()
+            if (a < 0) {
+                return
+            }
+        }
+
+        val p = if (alpha) 2 else 0
+        val r = parseCol(color, p + 0).toInt()
+        val g = parseCol(color, p + 2).toInt()
+        val b = parseCol(color, p + 4).toInt()
+        if (r < 0 || g < 0 || b < 0) {
+            return
+        }
+
+        this.r = r / 255.0
+        this.g = g / 255.0
+        this.b = b / 255.0
+        this.a = a / 255.0
+    }
+
+    constructor(from: Int) : this() {
+        TODO()
+    }
+
+    internal constructor(native: CValue<godot_color>) : this() {
+        memScoped {
+            this@Color.setRawMemory(native.ptr)
+        }
+    }
+
+    internal constructor(mem: COpaquePointer) : this() {
+        this.setRawMemory(mem)
+    }
+
+
+    //INTEROP
+    override fun getRawMemory(memScope: MemScope): COpaquePointer {
+        val value = cValue<godot_color_layout> {
+            r = this@Color.r.toFloat()
+            g = this@Color.g.toFloat()
+            b = this@Color.b.toFloat()
+            a = this@Color.a.toFloat()
+        }
+        return value.getPointer(memScope)
+    }
+
+    override fun setRawMemory(mem: COpaquePointer) {
+        val value = mem.reinterpret<godot_color_layout>().pointed
+        r = value.r.toRealT()
+        g = value.g.toRealT()
+        b = value.b.toRealT()
+        a = value.a.toRealT()
+    }
+
+    //API
+    /**
+     * Returns a new color resulting from blending this color over another.
+     * If the color is opaque, the result is also opaque. The second color may have a range of alpha values.
+     */
+    fun blend(over: Color): Color {
+        val res = Color()
+        val sa = 1.0 - over.a
+        if (res.a == 0.0) {
+            return Color(0.0, 0.0, 0.0, 0.0)
+        } else {
+            res.r = (r * a * sa + over.r * over.a) / res.a
+            res.g = (g * a * sa + over.g * over.a) / res.a
+            res.b = (b * a * sa + over.b * over.a) / res.a
+        }
+        return res
+    }
+
+    /**
+     * Returns the most contrasting color.
+     */
+    fun contrasted(): Color {
+        val c = this
```

Comment:
Here I'm not sure if we should change this color or provide a copy.

Reply:
I recreated the GDScript API:
[]( https://docs.godotengine.org/en/stable/classes/class_color.html?#class-color-method-color)

As all coreType are like primitive, all the methods return a modified copy of the original object. Unless you directly use the properties, you can't change the internal values of a non container coretype instance.

Reply:
This modifies the current object right?  Shouldn't it return a new copy?